### DISCUSSION TITLE
IndexError on pre-fitted SVC on FastGradientMethod

### DISCUSSION THREAD

**Discussion by ghost** [Trusted-AI/adversarial-robustness-toolbox#1237] [Q&A] [answered]
Hello,
I have a very simple question.
I have a basic setup with a trained SVC and I would like to try FastGradientMethod on it. I create SklearnClassifier with the following line based on my sklearn SVC:
classifier = SklearnClassifier(model=svc)

However I get "IndexError: arrays used as indices must be of integer (or boolean) type" when I try calling FastGradientMethod. 
**I do not want to re-fit my SklearnClassifier as my SVC is already fitted.** 

I would appreciate any help.
Thank you

**code snippet: (I can share more if needed)**
from art.attacks.evasion import FastGradientMethod
from art.estimators.classification import SklearnClassifier

classifier = SklearnClassifier(model=svc)
attack = FastGradientMethod(estimator=classifier, eps=0.2)
x_test_adv = attack.generate(x=scaledXTest)

**Error:**
IndexError                                Traceback (most recent call last)
<ipython-input-12-d97a812a8cdf> in <module>
      4 classifier = SklearnClassifier(model=svc)
      5 attack = FastGradientMethod(estimator=classifier, eps=0.2)
----> 6 x_test_adv = attack.generate(x=scaledXTest)

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/art/attacks/attack.py in replacement_function(self, *args, **kwargs)
     73                 if len(args) > 0:
     74                     args = tuple(lst)
---> 75                 return fdict[func_name](self, *args, **kwargs)
     76 
     77             replacement_function.__doc__ = fdict[func_name].__doc__

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/art/attacks/evasion/fast_gradient.py in generate(self, x, y, **kwargs)
    221                 # Use model predictions as correct outputs
    222                 logger.info("Using model predictions as correct labels for FGM.")
--> 223                 y = get_labels_np_array(self.estimator.predict(x, batch_size=self.batch_size))  # type: ignore
    224             y = y / np.sum(y, axis=1, keepdims=True)
    225 

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/art/estimators/classification/classifier.py in replacement_function(self, *args, **kwargs)
     69                 if len(args) > 0:
     70                     args = tuple(lst)
---> 71                 return fdict[func_name](self, *args, **kwargs)
     72 
     73             replacement_function.__doc__ = fdict[func_name].__doc__

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/art/estimators/classification/scikitlearn.py in predict(self, x, **kwargs)
   1507             y_pred_label = self.model.predict(X=x_preprocessed)
   1508             targets = np.array(y_pred_label).reshape(-1)
-> 1509             one_hot_targets = np.eye(self.nb_classes)[targets]
   1510             y_pred = one_hot_targets
   1511 

IndexError: arrays used as indices must be of integer (or boolean) type

**Comment by beat-buesser** [comment, marked as answer]
Hi @melikedem Thank you very much for using ART!
Could you please describe the shape of your input and label data? Is it binary or multi-class classification?

**Reply by ghost** [reply]
Thank you so much for your fast reply!
It is multi-class classification.
X: (3550, 73)
Y: (3550, 1)

**Reply by beat-buesser** [reply]
Thank you! I have not yet been able to reproduce it, but I'm interested to see why this happens. Would you be able to share a code example that I can run to reproduces the error? It can work with random example data in case you are working on private data.

**Reply by ghost** [reply]
Thank you again!
I attached my code. Yes I am working on my private data which is attached as well.
[Datasets.zip](https://github.com/Trusted-AI/adversarial-robustness-toolbox/files/6819051/Datasets.zip)
[fast gradient.ipynb.zip](https://github.com/Trusted-AI/adversarial-robustness-toolbox/files/6819053/fast.gradient.ipynb.zip)

**Reply by beat-buesser** [reply]
I have noticed two items in your notebook:
- `sklearn.svm.SVC.predict` returns the same `dtype` as has been used in `sklearn.svm.SVC.fit` and in line `svc.fit(scaledXTrain, y_train.values.ravel())` the `ravel()` returns indices of type `np.float64`. The current version of `SklearnClassifier` only supports labels of `dtype=np.int*`
- after fixing the first item, there is another issue, the labels in `y_train` are 1 to 14 for the 14 classes, but `art.estimators.classification` is expecting label indices starting from 0, therefore `y_train` should be 0 to 13

Changing `svc.fit(scaledXTrain, y_train.values.ravel())` to `svc.fit(scaledXTrain, y_train.values.ravel().astype(np.int)-1)` is one way of fixing both, but storing the new labels in a separate variable would be better.

One item we learned from your question is that we should think about that model predictions in `SklearnClassifier` don't have to be always `np.int` and that we probably should add a check and conversion for other dtypes.

**Reply by ghost** [reply]
Thank you so much for pointing out the problem.
If it is possible to support other data types with a simple check that would be great!
Have a nice day!
